plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+
+from django.core.files import File
+from django.core.files.storage import default_storage
+from django.db.models import Q
+
+from pulpcore.plugin.models import Artifact, ProgressBar
+
+
+async def query_existing_artifacts(in_q, out_q):
+    """
+    Stages API stage that replaces :attr:`DeclarativeContent.content` objects with already-saved
+    :class:`~pulpcore.plugin.models.Artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    This stage inspects any unsaved :class:`~pulpcore.plugin.models.Artifact` objects and searches
+    using their metadata for existing saved :class:`~pulpcore.plugin.models.Artifact` objects inside
+    Pulp with the same digest value(s). Any existing :class:`~pulpcore.plugin.models.Artifact`
+    objects found will replace their unsaved counterpart in the
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The query_existing_artifacts stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content = []
+    shutdown = False
+    while True:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content:
+                content = await in_q.get()
+                declarative_content.append(content)
+                continue
+        else:
+            declarative_content.append(content)
+            continue
```

Comment:
This *batching* logic seems to rely on a momentary gap in the `in_q` to determine the batch size.  If the `in_q` is being fed faster than consumed here, won't that result in all of the content being in memory at once.  Conversely, what if the `in_q` is being fed in slow, small bursts.  Won't that result in really small batch sizes and poor performance?

Reply:
It's possible that it could have small batch sizes, but it's also possible that adding delays could produce poor performance. I think we just don't know without data. Overall I think we need some performance tuning, but we need some fixture data to benchmark it against so we can see how our changes are affecting it. I think we should do these kinds of things later. The great thing is that each stage can be tuned and benchmarked individually.